### DISCUSSION TITLE
Adding use_terms_query for cardinality rules

### DISCUSSION THREAD

**Discussion by DrewHorrigan** [jertel/elastalert2#694] [Ideas]
My current use case is a rule to detect port scans, running on _a lot_ of traffic data. Currently we use a cardinality rule with a query_key containing 2 fields, but it downloads full events and counts the cardinality on the side of Elastalert, which makes the rule very heavy, sometimes to the point where it crashes or times out (I tried to use "include" to only retrieve the required fields, but this does not seem to help much).

This seems to be an appropriate use for a Terms aggregation; if query_key is used, it can be a series of embedded Terms aggregations, or, alternatively, a [Composite aggregation](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/search-aggregations-bucket-composite-aggregation.html). I tested the performance of the latter, and it seems to be faster than the full query by orders of magnitude (it also has an added advantage of a possibly unlimited number of terms). I've been wondering why it wasn't present in the original Elastalert - is there a technical reason preventing this?

**Comment by jertel** [comment]
[Metric Aggregations](https://elastalert2.readthedocs.io/en/latest/ruletypes.html?highlight=metric_aggregation#metric-aggregation) is available for similar purposes. Have you looked to see if this can meet your needs?

**Reply by DrewHorrigan** [reply]
I haven't, actually - I guess I overlooked it. Will definitely try it, thanks.

**Comment by JeffAshton** [comment]
Maybe related:

https://github.com/jertel/elastalert2/discussions/699
